refactor: update `stats/base/dists/triangular/mgf` implementation and test tolerances

anandkaranubc opened the pull request:
Resolves [#4704](https://github.com/stdlib-js/stdlib/issues/4704)

## Description

> What is the purpose of this pull request?

This pull request:

-  updates the implementation of the triangular MGF in `main.c`, `main.js`, and `factory.js` to align with the latest [Julia](https://github.com/JuliaStats/Distributions.jl/blob/957f0c0f79307edefb9afa75c682ef18b35dd982/src/univariate/continuous/triangular.jl#L139-L171) standards. It also updates test tolerances in `test.native.js` for consistency.
- update fixtures for triangular MGF based on Julia implementation (using the same Julia version 1.5)

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves [#4704](https://github.com/stdlib-js/stdlib/issues/4704)

## Questions

> Any questions for reviewers of this pull request?

- There is a helper function called [phi2](https://github.com/JuliaStats/Distributions.jl/blob/957f0c0f79307edefb9afa75c682ef18b35dd982/src/univariate/continuous/triangular.jl#L139-L151) which has been added to the C and JS files. I have added it as a private helper function in the same files. Should it be moved to a separate module for JS and a separate C file?

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

- The PR reduces the tolerance thresholds for `test.native.js` from a factor of `1.0e7` to `4.0`. The required tolerance should be `1.0`, but only 2 test cases fail with a `1.0` factor tolerance. I tried investigating the root cause by changing the order of terms and avoiding operator fusion, but no changes were observed. I will need some advice on how to fix this.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: update fixtures for triangular MGF based on Julia implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: align main and factory.js with Julia-based triangular MGF implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: adjust tolerance levels in test.native.js for triangular MGF consistency

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: update main.c to align triangular MGF implementation with Julia

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc committed:
refactor: reorganise unused variables in main.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
Hii @kgryte,

Over to you for review. A follow up question in addition to the ones above:

- To ensure 100% coverage, should I update `runner.jl` to include examples from all [4 cases](https://github.com/JuliaStats/Distributions.jl/blob/957f0c0f79307edefb9afa75c682ef18b35dd982/src/univariate/continuous/triangular.jl#L156-L169), or should I just manually add additional test cases in a new tape function?

kgryte commented:
@anandkaranubc Would you mind updating and rerunning the runner?

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/factory.js:
```diff
@@ -44,23 +61,16 @@ var pow = require( '@stdlib/math/base/special/pow' );
-	var bma;
-	var cma;
-
+function factory(a, b, c) {
```
```suggestion
function factory( a, b, c ) {
```

stdlib conventions are to use spaces within brackets. Applies here and elsewhere throughout this PR.

anandkaranubc committed:
feat: add fixtures for triangular MGF test cases for different conditions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add fixtures for triangular MGF test cases for different conditions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
bench: added benchmarks for different cases in triangular mgf implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc committed:
refactor: remove the extra if in main.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
chore: fix the order of includes in main.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
Can I get some feedback on this too please, @kgryte? Thanks!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/factory.js:
```diff
@@ -23,7 +23,24 @@
+* @returns {number} evaluated result
+*/
+function phi2( x ) {
+	if ( x === 0 ) {
+		return 1;
+	}
+	return ( 2 * ( expm1( x ) - x ) ) / ( x * x );
```
```suggestion
	if ( x === 0.0 ) {
		return 1.0;
	}
	return ( 2.0 * ( expm1( x ) - x ) ) / ( x * x );
```

`phi2` generally expects and returns doubles. As such, similar to C, we add decimals to indicate floating-point values.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/factory.js:
```diff
@@ -75,19 +85,24 @@ function factory( a, b, c ) {
-			return 1.0;
+		if ( a < c ) {
+			if ( c < b ) {
+				return (
+					exp( c * t ) * (
+						( ( c - a ) * phi2( ( a - c ) * t ) ) +
+						( ( b - c ) * phi2( ( b - c ) * t ) )
+					) / ( b - a )
+				);
```
```suggestion
				return exp( c*t ) * ( ( (c-a)*phi2( (a-c)*t ) ) + ( (b-c)*phi2( (b-c)*t ) ) ) / ( b-a ); // eslint-disable-line max-len
```

kgryte replied:
This is a minor quibble. In short, my preference is to use negative space to visually group related symbols and to draw the eye to the most salient information. This is philosophy underlying our use of negative space throughout the project.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/main.js:
```diff
@@ -22,7 +22,24 @@
+* @returns {number} evaluated result
+*/
+function phi2( x ) {
+	if ( x === 0 ) {
```
Same comments as other file.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/main.js:
```diff
@@ -88,17 +100,21 @@ function mgf( t, a, b, c ) {
+	if ( a < c ) {
+		if ( c < b ) {
+			return (
+				exp( c * t ) * (
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/factory.js:
```diff
@@ -23,7 +23,24 @@
+* @param {number} x - input value
+* @returns {number} evaluated result
+*/
+function phi2( x ) {
```
For the JS implementation, I suggest moving `phi2` to a separate file and then importing in both this file and in `main.js`. This will reduce code duplication.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/src/main.c:
```diff
@@ -18,8 +18,22 @@
+/**
+* Helper function for repeated computation in the MGF formula.
+*
+* @private
```
```suggestion
```

`static` already conveys "private".

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/src/main.c:
```diff
@@ -18,8 +18,22 @@
+* @param x   input value
+* @return    evaluated result
+*/
+static double phi2( const double x ) {
```
This, however, is fine and does **not** need to be split out into a separate file.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/test/test.factory.js:
```diff
@@ -224,3 +228,127 @@ tape( 'the created function evaluates the MGF for `x` given a large range `b - a
 	t.end();
 });
+
+tape( 'the function evaluates the MGF for `x` given the case: a < c < b', function test( t ) {
```
```suggestion
tape( 'the created function evaluates the MGF for `x` given the case: a < c < b', function test( t ) {
```

The change applies here and below and ensures consistency with the descriptions above.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/test/test.native.js:
```diff
@@ -165,9 +165,123 @@ tape( 'the function evaluates the MGF for `x` given a large range `b - a`', opts
+			t.equal( y, expected[i], 'x: '+x[i]+', a: '+a[i]+', b: '+b[i]+', c: '+c[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[ i ] );
+			tol = 4.0 * EPS * abs( expected[ i ] );
```
While muuuuch closer to JS, it isn't a perfect match. Any idea why?

kgryte replied:
After re-reading your OP, have you tried recompiling but disabling certain compiler optimizations?

anandkaranubc replied:
Update: I added `CFLAGS` to the `Makefile` to disable compiler optimizations (`-O0 -g`). This did impact the test results, with only one test failing after adjusting all tolerances from a factor of `4.0` to `1.0`.

I'll bring this up in today's OH to clarify if this is the best approach or if there's an alternative solution. I've also added a question to the OH issue. See you soon!

anandkaranubc replied:
Hi again @kgryte,  

Here's an update on the issue: I realized that I was considering higher tolerances for all cases, but in most cases, a factor of `1.0` works just fine.  

There is, however, one test that fails in this test category:  
https://github.com/stdlib-js/stdlib/blob/0937887902aed132ffe165cc713f053bb28fe369/lib/node_modules/%40stdlib/stats/base/dists/triangular/mgf/test/test.native.js#L88  

I reviewed the discussion in https://github.com/stdlib-js/stdlib/pull/2298, but it seems that using `CFLAGS="-ffp-contract=off"` or `CFLAGS="-ffp-contract=on"` doesn't make a difference in this case.  

I tracked down the first mismatch in the results between the JS and C implementations to the calculation of `expm1`:  
https://github.com/stdlib-js/stdlib/blob/0937887902aed132ffe165cc713f053bb28fe369/lib/node_modules/%40stdlib/stats/base/dists/triangular/mgf/src/main.c#L34  

The C implementation gives:  
`stdlib_base_expm1((a - c) * t)` & `stdlib_base_expm1((b - c) * t)`  
- term1: `-0.75243526099293723`  
- term2: `0.51382856876570271`  

The JS implementation gives:  
`expm1((a - c) * t)` & `expm1((b - c) * t)`  
- term1: `-0.7524352609929372`  
- term2: `0.5138285687657028`  

This doesn't seem to be the result of simple rounding errors like those seen in the other issue. As you mentioned, this is likely due to different implementations or approximations used for these functions. Any thoughts? :)

anandkaranubc replied:
Also, for the one test that fails, a factor of `2.0` instead of `1.0` works. Added a comment in the recent commit.

kgryte replied:
Adjusting to a tolerance of `2.0` should be fine. In the test comment, you can link to this discussion.

anandkaranubc replied:
Already done. Hope this looks good:

https://github.com/stdlib-js/stdlib/blob/2d4f12d50db7baa89db55d9f1fb9c20564efb9d4/lib/node_modules/%40stdlib/stats/base/dists/triangular/mgf/test/test.native.js#L111

anandkaranubc committed:
refactor: apply suggestions from PR review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc committed:
refactor: add private label to phi2.js

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
bench: update tolerances for the native tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc committed:
bench: move random number generation outside the benchmarking loops and use uniform instead of randu

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
Should the `README.md` also be changed with the updated formula for mgf calculation?

```math
M_X(t) := \mathbb{E}\!\left[e^{tX}\right] =
\begin{cases}
\exp(ct) \frac{(c-a)\phi_2((a-c)t) + (b-c)\phi_2((b-c)t)}{b-a} & \text{if } a < c < b \\
\exp(ct) \phi_2((a-c)t) & \text{if } a = c < b \\
\exp(ct) \phi_2((b-c)t) & \text{if } a < c = b \\
\exp(ct) & \text{if } a = c = b
\end{cases}
```

```math
\phi_2(x) :=
\begin{cases}
1, & \text{if } x = 0 \\
\frac{2 \left( \exp(x) - 1 - x \right)}{x^2} & \text{if } x \neq 0
\end{cases}
```

I notice that this would also mean changing the svg link provided for the equation.

A side note:  The Julia implementation does this case by case breakdown to avoid unnecessary computation.

kgryte commented:
> Should the README.md also be changed with the updated formula for mgf calculation?

No. The equations in READMEs should not necessarily reflect implementation logic. That is an implementation detail. Unless the package explicitly concerns a particular algorithm, the README contains equations as you'd find in a textbook or Wikipedia. The MGF currently in the README follows the latter.

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into fix/triangular-mgf-divergence

anandkaranubc committed:
test: update comment about high tolerance issue

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc commented:
Once https://github.com/stdlib-js/stdlib/pull/5638 gets merged, the code coverage will be 100%.

Edit: That PR has been merged and the code coverage is 100% now!

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/test/test.native.js:
```diff
@@ -104,8 +108,8 @@ tape( 'the function evaluates the MGF for `x` given a small range `b - a`', opts
-			// FIXME: the tolerance for the C implementation is widely different from the JavaScript implementation, and it is not clear why.
-			tol = 1.0e7 * EPS * abs( expected[ i ] );
+			// NOTE: the tolerance here is larger than for the JavaScript implementation due to compiler optimizations which may be performed resulting in result divergence. For discussion, see https://github.com/stdlib-js/stdlib/pull/2298#discussion_r1624765205
+			tol = 2.0 * EPS * abs( expected[ i ] );
```
I can confirm that if `CFLAGS="-ffp-contract=off"` is used, the tolerance can be set to `1.0` here.

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix/triangular-mgf-divergence

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/mgf/lib/phi2.js:
```diff
@@ -0,0 +1,42 @@
+		return 1.0;
+	}
+	return ( 2.0 * ( expm1( x ) - x ) ) / ( x*x );
+}
+
+module.exports = phi2;
```
```suggestion
}


// EXPORTS //

module.exports = phi2;
```

Planeshifter committed:
chore: add exports header comment

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Let's get this one in! 🚀
